File: ios/Crew/Engine/PauseValidation.swift
// SPEC: V22 (never retroactive) · V23 (≤ pauseMaxDays) · V44 (one active pause at a time), in that order.
// Twin of pause-validation.ts. WRITTEN — UNVERIFIED (needs Mac).

import Foundation

enum PauseRejection: String, Codable {
    case retroactive, tooLong, alreadyPaused
}

struct PauseValidationResult: Equatable {
    let accepted: Bool
    let reason: PauseRejection?
}

enum PauseValidation {
    static func validatePauseRequest(today: String, startDay: String, endDay: String, existingPauses: [Pause]) -> PauseValidationResult {
        if startDay < today { return PauseValidationResult(accepted: false, reason: .retroactive) }
        if DayKey.daysBetween(startDay, endDay) > SpecConstants.pauseMaxDays { return PauseValidationResult(accepted: false, reason: .tooLong) }
        if existingPauses.contains(where: { $0.endDay > today }) { return PauseValidationResult(accepted: false, reason: .alreadyPaused) }
        return PauseValidationResult(accepted: true, reason: nil)
    }
}

File: web/src/lib/engine/pause-validation.ts
// SPEC: V22 (never retroactive) · V23 (≤ pauseMaxDays) · V44 (one active pause at a time), checked in that order.
// Twin: ios/Crew/Engine/PauseValidation.swift.
import { daysBetween } from "@/lib/engine/day-key";
import type { Pause } from "@/lib/engine/gamification";
import { SpecConstants } from "@/generated/spec-constants";

export type PauseRejection = "retroactive" | "tooLong" | "alreadyPaused";

export interface PauseValidation {
  accepted: boolean;
  reason?: PauseRejection;
}

export function validatePauseRequest(today: string, startDay: string, endDay: string, existingPauses: Pause[]): PauseValidation {
  if (startDay < today) return { accepted: false, reason: "retroactive" };
  if (daysBetween(startDay, endDay) > SpecConstants.pauseMaxDays) return { accepted: false, reason: "tooLong" };
  if (existingPauses.some((pause) => pause.endDay > today)) return { accepted: false, reason: "alreadyPaused" };
  return { accepted: true };
}
